### PR TITLE
 Add an API to retrieve MMIO region information per partition

### PR DESCRIPTION
**Pull request by romkuz01**
- Generate a global table of all the MMIO regions.
- Add an API get_mmio_regions() to fetch MMIO regions for a given partition (or all the partitions at once).

---
This change is [Reviewable](https://reviewable.io/reviews/kfnta/mbed-os/56)

### CODE REVIEW

File: tools/spm/templates/psa_common.c.tpl [hunk 1]
```diff
@@ -75,12 +75,25 @@ spm_partition_t *g_partitions = NULL;
...
 /* Check all the defined MMIO regions for overlapping. */
 {% for region_pair in region_pair_list %}
-static_assert(
+_Static_assert(
     ((uintptr_t)({{region_pair[0].base}}) + {{region_pair[0].size}} - 1 < (uintptr_t)({{region_pair[1].base}})) ||
     ((uintptr_t)({{region_pair[1].base}}) + {{region_pair[1].size}} - 1 < (uintptr_t)({{region_pair[0].base}})),
     "The region with base {{region_pair[0].base}} and size {{region_pair[0].size}} overlaps with the region with base {{region_pair[1].base}} and size {{region_pair[1].size}}!");
 {% endfor %}
 
+/* A list of all the MMIO regions. */
```
Comment (orenc17) [inline tools/spm/templates/psa_common.c.tpl]: in my opinion mmio should be inside the partition struct
so this whole decalartion should sit in psa_NAME_partition.c.tpl

Comment (romkuz01) [reply tools/spm/templates/psa_common.c.tpl]: The idea was to have an easy way to retrieve a list for all the partitions together at once, thus global array is convenient for this purpose.

File: spm/spm_common.c [hunk 2]
```diff
@@ -69,3 +69,30 @@ inline void channel_state_assert(uint8_t *current_state, uint8_t expected_state)
...
     }
 }
+
+extern const mmio_region_t *mmio_regions;
+extern const uint32_t mmio_region_count;
+
+const mmio_region_t *get_mmio_regions(int32_t partition_id, uint32_t *region_count)
+{
+    uint32_t i;
+    const mmio_region_t *regions;
```
Comment (danny4478) [inline spm/spm_common.c]: Can be declared when used

File: spm/spm_common.c [hunk 3]
```diff
@@ -69,3 +69,30 @@ inline void channel_state_assert(uint8_t *current_state, uint8_t expected_state)
             *current_state, expected_state);
     }
 }
+
+extern const mmio_region_t *mmio_regions;
+extern const uint32_t mmio_region_count;
+
+const mmio_region_t *get_mmio_regions(int32_t partition_id, uint32_t *region_count)
+{
+    uint32_t i;
```
Comment (danny4478) [inline spm/spm_common.c]: I think we should assert that (region_count != NULL)

File: spm/spm_common.c [hunk 4]
```diff
@@ -69,3 +69,30 @@ inline void channel_state_assert(uint8_t *current_state, uint8_t expected_state)
...
+    }
+
+    for (i = 0; i < mmio_region_count && mmio_regions[i].partition_id != partition_id; i++);
+
+    if (i == mmio_region_count) {
+        return NULL;
+    }
+
+    regions = &mmio_regions[i];
+    for (; i < mmio_region_count && mmio_regions[i].partition_id == partition_id; i++, *region_count++);
```
Comment (danny4478) [inline spm/spm_common.c]: Not sure what the precedence is. 
I would use parentheses anyway: (*region_count)++

File: spm/spm_hal_sw.c [hunk 5]
```diff
@@ -40,3 +40,8 @@ void nspe_done(osSemaphoreId_t completion_sem_id)
     SPM_ASSERT(osOK == os_status);
     PSA_UNUSED(os_status);
 }
+
+void memory_protection_init(const mem_region_t *regions, uint32_t region_count)
```
Comment (orenc17) [inline spm/spm_hal_sw.c]: i think this should be implemented only in platforms that allow memory protection
we should not allow an empty implementation

File: spm/spm_init.c [hunk 6]
```diff
@@ -72,6 +72,10 @@ uint32_t init_partitions(spm_partition_t **partitions);
 
 void psa_spm_init(void)
 {
+    uint32_t region_count = 0;
```
Comment (orenc17) [inline spm/spm_init.c]: surround this with an \#ifdef SPM_MEMORY_PROTECTION_ENABLED

File: spm/spm_hal_sw.c (hunk 5 above)
Comment (alzix) [reply spm/spm_hal_sw.c]: the way i see it - we have to have per implementation source file. that is why we have this HALfile in the first place.
defining it week is a bad idea.
We will have to provide  different implementation for other platform. Linker will make sure you are not forgetting to implement it. That is why weak symbols are bad idea. compile time polymorphism.

File: spm/spm_init.c (hunk 6 above)
Comment (alzix) [reply spm/spm_init.c]: disagree

**Pull request merged**